File: src/index/topk.ts
export interface TopKDistanceEntry {
  index: number;
  distance: number;
}

/** Keep the k smallest distances (best cosine matches). Legacy reference implementation. */
export function selectTopKByDistance(
  entries: Iterable<TopKDistanceEntry>,
  k: number,
): TopKDistanceEntry[] {
  const top: TopKDistanceEntry[] = [];
  for (const entry of entries) {
    if (top.length < k) {
      top.push(entry);
      top.sort((a, b) => b.distance - a.distance);
      continue;
    }
    const worst = top[0];
    if (worst && entry.distance < worst.distance) {
      top[0] = entry;
      top.sort((a, b) => b.distance - a.distance);
    }
  }
  top.sort((a, b) => a.distance - b.distance);
  return top;
}

/**
 * Incremental top-k by smallest distance using a size-k max-heap.
 * Avoids materializing all candidate scores before selection.
 */
export class TopKDistanceCollector {
  private readonly heap: TopKDistanceEntry[] = [];

  constructor(private readonly k: number) {}

  offer(index: number, distance: number): void {
    if (this.k <= 0) {
      return;
    }
    if (this.heap.length < this.k) {
      this.heap.push({ index, distance });
      this.bubbleUp(this.heap.length - 1);
      return;
    }
    const worst = this.heap[0];
    if (worst && distance < worst.distance) {
      this.heap[0] = { index, distance };
      this.siftDown(0);
    }
  }

  finish(): TopKDistanceEntry[] {
    return [...this.heap].sort((a, b) => a.distance - b.distance);
  }

  private parent(i: number): number {
    return (i - 1) >> 1;
  }

  private left(i: number): number {
    return (i << 1) + 1;
  }

  private bubbleUp(i: number): void {
    while (i > 0) {
      const p = this.parent(i);
      const here = this.heap[i];
      const parent = this.heap[p];
      if (!here || !parent || here.distance <= parent.distance) {
        break;
      }
      this.heap[i] = parent;
      this.heap[p] = here;
      i = p;
    }
  }

  private siftDown(i: number): void {
    while (true) {
      const left = this.left(i);
      if (left >= this.heap.length) {
        break;
      }
      const right = left + 1;
      let largest = left;
      if (
        right < this.heap.length &&
        (this.heap[right]?.distance ?? -Infinity) > (this.heap[left]?.distance ?? -Infinity)
      ) {
        largest = right;
      }
      const here = this.heap[i];
      const child = this.heap[largest];
      if (!here || !child || child.distance <= here.distance) {
        break;
      }
      this.heap[i] = child;
      this.heap[largest] = here;
      i = largest;
    }
  }
}

/** Indices of the k highest scores in a dense array. */
export function selectTopKScoreIndices(scores: number[], k: number): number[] {
  if (k <= 0 || scores.length === 0) {
    return [];
  }
  const top: { v: number; i: number }[] = [];
  for (let i = 0; i < scores.length; i++) {
    const v = scores[i];
    if (v === undefined) {
      continue;
    }
    if (top.length < k) {
      top.push({ v, i });
      top.sort((a, b) => a.v - b.v);
      continue;
    }
    const worst = top[0];
    if (worst && v > worst.v) {
      top[0] = { v, i };
      top.sort((a, b) => a.v - b.v);
    }
  }
  top.sort((a, b) => b.v - a.v);
  return top.map((x) => x.i);
}

File: internal/index/topk.go
package index

import "sort"

type TopKDistanceEntry struct {
	Index    int
	Distance float64
}

func SelectTopKByDistance(entries []TopKDistanceEntry, k int) []TopKDistanceEntry {
	top := []TopKDistanceEntry{}
	for _, entry := range entries {
		if len(top) < k {
			top = append(top, entry)
			sort.Slice(top, func(i, j int) bool { return top[i].Distance > top[j].Distance })
			continue
		}
		if len(top) > 0 && entry.Distance < top[0].Distance {
			top[0] = entry
			sort.Slice(top, func(i, j int) bool { return top[i].Distance > top[j].Distance })
		}
	}
	sort.Slice(top, func(i, j int) bool { return top[i].Distance < top[j].Distance })
	return top
}

func SelectTopKScoreIndices(scores []float64, k int) []int {
	if k <= 0 || len(scores) == 0 {
		return nil
	}

	type scoreEntry struct {
		value float64
		index int
	}
	top := []scoreEntry{}
	for i, score := range scores {
		if len(top) < k {
			top = append(top, scoreEntry{value: score, index: i})
			sort.Slice(top, func(i, j int) bool { return top[i].value < top[j].value })
			continue
		}
		if len(top) > 0 && score > top[0].value {
			top[0] = scoreEntry{value: score, index: i}
			sort.Slice(top, func(i, j int) bool { return top[i].value < top[j].value })
		}
	}
	sort.Slice(top, func(i, j int) bool { return top[i].value > top[j].value })
	out := make([]int, len(top))
	for i, entry := range top {
		out[i] = entry.index
	}
	return out
}
